### PR TITLE
Supporting erc20

### PR DESCRIPTION
**Pull request by JIUQIWANG**
Working on #1243 

Add 6 ERC20 standard command-line function to ioctl:

ioctl xrc20 totalsupply -c ALIAS|CONTRACT_ADDRESS

ioctl xrc20 balanceof -c ALIAS|CONTRACT_ADDRESS (ALIAS|OWNER_ADDRESS)

ioctl xrc20 transfer (ALIAS|TARGET_ADDRESS) (AMOUNT) -c ALIAS|CONTRACT_ADDRESS -l GAS_LIMIT -s SIGNER [-p GAS_PRICE] 

ioctl xrc20 transferfrom (ALIAS|OWNER_ADDRESS) (ALIAS|TARGET_ADDRESS) (AMOUNT) -c (ALIAS|CONTRACT_ADDRESS) -l GAS_LIMIT -s SIGNER [-p GAS_PRICE]

ioctl xrc20 approve (ALIAS|SPENDER_ADDRESS) (ALIAS|AMOUNT) -c ALIAS|CONTRACT_ADDRESS -s SIGNER -l GAS_LIMIT 

ioctl xrc20 allowance (ALIAS|OWNER_ADDRESS) (ALIAS|SPENDER_ADDRESS) -c ALIAS|CONTRACT_ADDRESS -s SIGNER 

I test them on the testnet endpoint:
using following contract:

io1y9ndaezjrdlkw93hquqru7txh9jcsmtmrvt4yw

And two accounts under this contract:

io1juvx5g063eu4ts832nukp4vgcwk2gnc5cu9ayd

io1q4enhh0tp5pqpa6s4urhwrx32529pmyyzdgu3q 

And test all six functions, they all works. I am still working on some small issue to make code more simplify and more readable.

### CODE REVIEW

File: cli/ioctl/cmd/xrc20/xrc20.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
...nce\",\r\n\t\t\"outputs\": [\r\n\t\t\t{\r\n\t\t\t\t\"name\": \"\",\r\n\t\t\t\t\"type\": \"uint256\"\r\n\t\t\t}\r\n\t\t],\r\n\t\t\"payable\": false,\r\n\t\t\"stateMutability\": \"view\",\r\n\t\t\"type\": \"function\"\r\n\t},\r\n\t{\r\n\t\t\"anonymous\": false,\r\n\t\t\"inputs\": [\r\n\t\t\t{\r\n\t\t\t\t\"indexed\": true,\r\n\t\t\t\t\"name\": \"owner\",\r\n\t\t\t\t\"type\": \"address\"\r\n\t\t\t},\r\n\t\t\t{\r\n\t\t\t\t\"indexed\": true,\r\n\t\t\t\t\"name\": \"spender\",\r\n\t\t\t\t\"type\": \"address\"\r\n\t\t\t},\r\n\t\t\t{\r\n\t\t\t\t\"indexed\": false,\r\n\t\t\t\t\"name\": \"value\",\r\n\t\t\t\t\"type\": \"uint256\"\r\n\t\t\t}\r\n\t\t],\r\n\t\t\"name\": \"Approval\",\r\n\t\t\"type\": \"event\"\r\n\t},\r\n\t{\r\n\t\t\"anonymous\": false,\r\n\t\t\"inputs\": [\r\n\t\t\t{\r\n\t\t\t\t\"indexed\": true,\r\n\t\t\t\t\"name\": \"from\",\r\n\t\t\t\t\"type\": \"address\"\r\n\t\t\t},\r\n\t\t\t{\r\n\t\t\t\t\"indexed\": true,\r\n\t\t\t\t\"name\": \"to\",\r\n\t\t\t\t\"type\": \"address\"\r\n\t\t\t},\r\n\t\t\t{\r\n\t\t\t\t\"indexed\": false,\r\n\t\t\t\t\"name\": \"value\",\r\n\t\t\t\t\"type\": \"uint256\"\r\n\t\t\t}\r\n\t\t],\r\n\t\t\"name\": \"Transfer\",\r\n\t\t\"type\": \"event\"\r\n\t}\r\n]"
```
Comment (CoderZhi) [inline cli/ioctl/cmd/xrc20/xrc20.go]: use ``: https://golang.org/ref/spec#String_literals

File: cli/ioctl/cmd/xrc20/xrc20.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2019 IoTeX
```
Comment (CoderZhi) [inline cli/ioctl/cmd/xrc20/xrc20.go]: This command could be placed in https://github.com/iotexproject/iotex-core/blob/master/cli/ioctl/cmd/action/action.go
then, you can refer to this command as action.Xrc20Cmd

File: cli/ioctl/cmd/root.go [hunk 3]
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/iotexproject/iotex-core/cli/ioctl/cmd/node"
 	"github.com/iotexproject/iotex-core/cli/ioctl/cmd/update"
 	"github.com/iotexproject/iotex-core/cli/ioctl/cmd/version"
+	xrc20 "github.com/iotexproject/iotex-core/cli/ioctl/cmd/xrc20"
```
[the remark this answers is not in the document]
Comment (JIUQIWANG) [reply cli/ioctl/cmd/root.go]: Thanks for all the comments! I will have modification today

File: cli/ioctl/cmd/action/action.go [hunk 4]
```diff
@@ -33,11 +35,18 @@ import (
...
+	gasPrice        string
+	nonce           uint64
+	signer          string
+	bytecodeString  string
+	contractAddress string
+	ownerAddress    address.Address
+	transferAmount  uint64
+	targetAddress   address.Address
+	spenderAddress  address.Address
+	abiResult       abi.ABI
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/action.go]: rename to xrc20ABI

File: cli/ioctl/cmd/action/action.go [hunk 5]
```diff
@@ -33,11 +35,18 @@ import (
...
+	bytes           []byte
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/action.go]: bad naming

File: cli/ioctl/cmd/action/action.go [hunk 6]
```diff
@@ -33,11 +35,18 @@ import (
...
-	gasPrice       string
-	nonce          uint64
-	signer         string
-	bytecodeString string
+	gasLimit        uint64
+	gasPrice        string
+	nonce           uint64
+	signer          string
+	bytecodeString  string
+	contractAddress string
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/action.go]: all the xrc20 only parameters should have prefix xrc20

File: cli/ioctl/cmd/action/action.go [hunk 7]
```diff
@@ -46,7 +55,25 @@ var ActionCmd = &cobra.Command{
...
 }
 
+//Xrc20Cmd represent erc20 standard command-line
+var Xrc20Cmd = &cobra.Command{
+	Use:   "xrc20",
+	Short: "Supporting ERC20 standard command-line from ioctl",
+}
+
 func init() {
+	abiJSON := abiConst
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/action.go]: delete abiJSON

File: cli/ioctl/cmd/action/action.go [hunk 8]
```diff
@@ -111,6 +145,19 @@ func setActionFlags(cmds ...*cobra.Command) {
 	}
 }
 
+func setActionFlagsXrc(cmds ...*cobra.Command) {
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/action.go]: merge with setActionFlagsAction, and name it as setFlags

File: cli/ioctl/cmd/action/xrc20allowance.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+		return err
+	},
+}
+
+// read reads smart contract on IoTeX blockchain
+func allowance(args []string) (string, error) {
+	var err error
+	args[0] = contractAddress
+	gasLimit = 50000
+	signer = "ALIAS"
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/xrc20allowance.go]: what's ALIAS?

File: cli/ioctl/cmd/action/xrc20approve.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+		}
+		transfer, err := strconv.ParseInt(args[1], 10, 64)
+		if err != nil {
+			return err
+		}
+		transferAmount = uint64(transfer)
+		output, err := approve(args)
+		if err == nil {
+			fmt.Println(output)
+
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/xrc20approve.go]: delete empty line

File: cli/ioctl/cmd/action/xrc20balanceof.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	},
+}
+
+func toEthAddr(addr address.Address) common.Address {
+	return common.BytesToAddress(addr.Bytes())
+}
+
+// read reads smart contract on IoTeX blockchain
+func balanceOf(args []string) (string, error) {
+
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/xrc20balanceof.go]: delete empty line

File: cli/ioctl/cmd/action/xrc20balanceof.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+		}
+		ownerAddress, err = address.FromString(addr)
+		if err != nil {
+			return err
+		}
+		output, err := balanceOf(args)
+		if err == nil {
+			fmt.Println(output)
+			result, _ := strconv.ParseUint(output, 16, 64)
+			fmt.Println("Output in decimal format:")
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/xrc20balanceof.go]: ```fmt.Printf("Ouptut in decimal format: %d", result)```

File: cli/ioctl/cmd/action/xrc20balanceof.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+			return err
+		}
+		ownerAddress, err = address.FromString(addr)
+		if err != nil {
+			return err
+		}
+		output, err := balanceOf(args)
+		if err == nil {
+			fmt.Println(output)
+			result, _ := strconv.ParseUint(output, 16, 64)
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/xrc20balanceof.go]: you cannot omit the error

Comment (CoderZhi) [inline cli/ioctl/cmd/action/xrc20balanceof.go]: uint64 may not be big enough, use big.Int

File: cli/ioctl/cmd/action/xrc20allowance.go [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	args[0] = xrc20ContractAddress
+	gasLimit = 50000
+	signer = "io1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqd39ym7"
+	xrc20Bytes, err = xrc20ABI.Pack("allowance", toEthAddr(xrc20OwnerAddress), toEthAddr(xrc20SpenderAddress))
+	if err != nil {
+		log.L().Error("cannot generate bytecode from given command", zap.Error(err))
+		return "", err
+	}
+	return read(args)
+
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/xrc20allowance.go]: delete empty line

File: cli/ioctl/cmd/action/xrc20approve.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+		}
+		xrc20SpenderAddress, err = address.FromString(addr)
+		if err != nil {
+			return err
+		}
+		transfer, err := strconv.ParseInt(args[1], 10, 64)
+		if err != nil {
+			return err
+		}
+		xrc20TransferAmount = uint64(transfer)
```
Comment (CoderZhi) [inline cli/ioctl/cmd/action/xrc20approve.go]: bigInt

**Pull request merged**